### PR TITLE
dependency: fix merging tags in 'merge_repeats'

### PR DESCRIPTION
**Pull request by UniqMartin**
While it may suffice to merge string and non-reserved tags by forming a union of all tags of dependencies of the same name, this approach fails to work for the reserved tags. These are now merged such that the most restrictive tag (meaning sometimes an empty tag) is preserved.

The previous behavior caused essential dependencies to be omitted and builds to fail in response. E.g., multiple `:fortran` dependencies with tags `[]`, `[:recommended]`, and `[:optional]` would have been expanded and merged to `"gcc"` with tags `[:recommended, :optional]`, causing it to be no longer seen as a required dependency.

To reproduce the issue fixed by this PR, start from a clean Homebrew installation and then execute:

```
$ brew tap homebrew/science
$ brew install -v petsc
==> Installing petsc from homebrew/science
==> Installing dependencies for homebrew/science/petsc: sphinx-doc, cmake, superlu43, metis, parmetis, superlu_dist, veclibfort, scalapack, mumps, hypre, sundials, szip, hdf5, pkg-config, hwloc, tbb, suite-sparse, netcdf, fftw
[…]
```

Notice how the list of dependencies to be installed neither contains `gcc` nor `open-mpi`, though [`petsc`](https://github.com/Homebrew/homebrew-science/blob/16c758b2298186330183520e4882dbc173528e3b/petsc.rb#L22-L23) has a mandatory dependency on both due to its `:fortran` and `:mpi` requirement. With this PR applied, the same invocation of `brew install` instead correctly yields:

```
==> Installing petsc from homebrew/science
==> Installing dependencies for homebrew/science/petsc: xz, gmp, mpfr, libmpc, isl, gcc, pkg-config, makedepend, openssl, libevent, open-mpi, sphinx-doc, cmake, superlu43, metis, parmetis, superlu_dist, veclibfort, scalapack, mumps, hypre, sundials, szip, hdf5, hwloc, tbb, suite-sparse, netcdf, fftw
[…]
```

**Notes:** I'm not particularly attached to any newly introduced names, I just couldn't come up with something better than “necessity” and “temporality”. I'd appreciate feedback for better naming. Please also scrutinize the merging logic and its representation in the tests.

(In some sense, this is a follow-up and extension of the fix in #46916.)

### CODE REVIEW

File: Library/Homebrew/test/test_dependency.rb [hunk 1]
```diff
@@ -48,7 +48,7 @@ def test_merge_repeats
     assert_equal Dependency, merged.first.class
 
     foo_named_dep = merged.find {|d| d.name == "foo"}
-    assert_equal [:build, "bar"], foo_named_dep.tags
+    assert_equal ["bar"], foo_named_dep.tags
```
Comment (mikemcquaid) [inline Library/Homebrew/test/test_dependency.rb]: This definitely seems wrong to me but I'll read more to understand why.

File: Library/Homebrew/dependency.rb [hunk 2]
```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
...
+
+    private
+
+    def merge_tags(deps)
+      options = deps.flat_map(&:option_tags).uniq
+      merge_necessity(deps) + merge_temporality(deps) + options
+    end
+
+    def merge_necessity(deps)
+      if deps.any? { |dep| !dep.recommended? && !dep.optional? }
```
Comment (mikemcquaid) [inline Library/Homebrew/dependency.rb]: I think `none?` might make more sense here.

File: Library/Homebrew/test/test_dependency.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Homebrew/test/test_dependency.rb]: Nope, I agree with the overall changes but this seems to be a mistake/regression.

File: Library/Homebrew/dependency.rb [hunk 3]
```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
...
+      end
+    end
+
+    def merge_temporality(deps)
+      if deps.all?(&:build?)
+        [:build]
+      elsif deps.all?(&:run?)
+        [:run]
+      else
+        [] # Implies build & runtime dependency.
```
Comment (mikemcquaid) [inline Library/Homebrew/dependency.rb]: Mind using `and` instead of `&` just to avoid confusion with the actual `&` operator?

File: Library/Homebrew/dependency.rb [hunk 4]
```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
...
+    private
+
+    def merge_tags(deps)
+      options = deps.flat_map(&:option_tags).uniq
+      merge_necessity(deps) + merge_temporality(deps) + options
+    end
+
+    def merge_necessity(deps)
+      if deps.any? { |dep| !dep.recommended? && !dep.optional? }
+        # FIXME: Should be `required?`, but `required?` implies `!build?`.
```
Comment (mikemcquaid) [inline Library/Homebrew/dependency.rb]: Not sure about the `FIXME` here given it doesn't actually apply to this code; maybe just make this a comment and add the `fixme` elsewhere.

**Comment by mikemcquaid** [comment]
Nice work here :+1:

File: Library/Homebrew/test/test_dependency.rb (hunk 1 above)
Comment (UniqMartin) [inline Library/Homebrew/test/test_dependency.rb]: I'm pretty sure the previous behavior is wrong. If we have a dependency tree that contains two instances of the same dependency, e.g.:

```
a
|-- xz => :build
\-- b
    \-- xz
```

Then `xz` is ultimately required at both build and runtime and thus the two instances of this dependency should be merged to the equivalent of `depends_on "xz"`.

File: Library/Homebrew/dependency.rb (hunk 3 above)
Comment (UniqMartin) [inline Library/Homebrew/dependency.rb]: If this helps avoid confusion, sure.

File: Library/Homebrew/dependency.rb (hunk 2 above)
Comment (UniqMartin) [inline Library/Homebrew/dependency.rb]: It's kinda related to the `FIXME` comment below. Ideally, I would be just using `deps.any?(&:required?)` instead of the block here, but because of the (debatable) definition of `required?` I'm doing it like this (and trying to keep it as close as possible to the definition of `required?`).

File: Library/Homebrew/dependency.rb (hunk 4 above)
Comment (UniqMartin) [inline Library/Homebrew/dependency.rb]: Yes, maybe the `FIXME` should go into the definition of `required?`.

Ideally, our treatment of dependencies would be a bit more uniform than is currently the case (if my assessment of the current situation is correct), resulting in a different definition of `required?` or both `optional?` and `recommended?` in `Dependable`. But that would be a too far-reaching change for this PR (and it requires substantially more testing). This PR on the other hand tries to be a fix that can be merged (hopefully) relatively quickly.

File: Library/Homebrew/test/test_dependency.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Homebrew/test/test_dependency.rb]: Of course! I see now, thanks for explaining.

File: Library/Homebrew/dependency.rb (hunk 4 above)
Comment (mikemcquaid) [inline Library/Homebrew/dependency.rb]: > Yes, maybe the FIXME should go into the definition of required?.

Yeh, I think so.

**Comment by mikemcquaid** [comment]
Other than the cosmetic changes here I'm :+1: to :ship: (although I suspect @xu-cheng may also want to take a look).

File: Library/Homebrew/dependency.rb [hunk 5]
```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
...
+        [] # Implies required dependency.
+      elsif deps.any?(&:recommended?)
+        [:recommended]
+      else # deps.all?(&:optional?)
+        [:optional]
+      end
```
Comment (xu-cheng) [inline Library/Homebrew/dependency.rb]: I wonder whether it would help to make the code more readable by implementing this function in this way:
```ruby
def merge_necessity(deps)
  recommended = false
  deps.each do |dep|
    if dep.recommended?
      recommended = true
    elsif !dep.optional?
     return []
    end
  end
  recommended ? [:recommended] : [:optional]
end
```

As a bonus, it will only enumerate the list for once.

File: Library/Homebrew/dependable.rb [hunk 6]
```diff
@@ -23,7 +23,15 @@ def required?
     !build? && !optional? && !recommended?
   end
 
+  def build_and_run?
+    !build? && !run?
+  end
```
Comment (xu-cheng) [inline Library/Homebrew/dependable.rb]: This method doesn't seem to be used anywhere except in test case. Should we keep it?

**Comment by xu-cheng** [comment]
Few small advice. LGTM :+1:

 :heart: to see we are addressing more edge cases .Thanks @UniqMartin

File: Library/Homebrew/dependency.rb (hunk 5 above)
Comment (UniqMartin) [inline Library/Homebrew/dependency.rb]: I personally think that my implementation is more readable, because it is expressed on a higher level. I find it easier to reason about all dependencies at once instead of understanding the various state changes that happen in the “loop” you suggest. And I think correctness is (in this case) more important than performance.

Here's my implementation translated into English which hopefully illustrates my point:

- If **any** of the dependencies is a **required** dependency:
  - The merged dependency shall be **required**.
- Otherwise, if **any** of the dependencies is a **recommended** dependency:
  - The merged dependency shall be **recommended**.
- Otherwise, **all** dependencies must be **optional** and:
  - The merged dependency shall be **optional**, too.

File: Library/Homebrew/dependable.rb (hunk 6 above)
Comment (UniqMartin) [inline Library/Homebrew/dependable.rb]: I agree that it is somewhat redundant. I removed it.

File: Library/Homebrew/dependency.rb (hunk 5 above)
Comment (mikemcquaid) [inline Library/Homebrew/dependency.rb]: Although @xu-cheng's involves a single loop I personally find @UniqMartin's more readable. I'm not convinced that the performance overhead is going to me enough to justify the drop in readability here.

**Comment by UniqMartin** [comment]
@mikemcquaid @xu-cheng I hope I addressed all of your comments either in the updated code I just pushed or by giving reasons why I'd prefer to stay with my implementation.

The issue solved here seems to affect more formulae than I expected – especially in homebrew/science, see <https://github.com/Homebrew/homebrew/pull/46461#issuecomment-165128667> –, so I'm tempted to merge this very soon.
